src/main/java/captain/cybot/adventure/backend/service/UserServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+import captain.cybot.adventure.backend.validator.PasswordConstraintValidator;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import javax.transaction.Transactional;
+import java.util.ArrayList;
+import java.util.Collection;
+
+@Service
+@AllArgsConstructor
+@Transactional
+@Slf4j
+public class UserServiceImpl implements UserService, UserDetailsService {
+
+    private UserRepository userRepository;
+    private RoleRepository roleRepository;
+    private CosmeticRepository cosmeticRepository;
+
+    @Autowired
+    public BCryptPasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
+
+    @Autowired
+    public PasswordConstraintValidator passwordConstraintValidator() {
+        return new PasswordConstraintValidator();
+    }
+
+    @Override
+    public User saveUser(User user) throws UserAlreadyExistsException, InvalidRoleException, PasswordInvalidException {
+        if(usernameAlreadyExists(user.getUsername())) {
+            throw new UserAlreadyExistsException("Username already exists.");
+        } else if(emailAlreadyExists(user.getEmail())) {
+            throw new UserAlreadyExistsException("Email already exists.");
+        } else if (!passwordConstraintValidator().isValid(user.getPassword())) {
+            throw new PasswordInvalidException("Password must contain the following: \n-8 to 24 characters" +
+                    "\n-ONE lowercase character \n-ONE uppercase character" +
+                    "\n-ONE digit character \n-ONE special character");
+        }
+        user.setPassword(passwordEncoder().encode(user.getPassword()));
+        User newUser = userRepository.save(user);
+        log.info(newUser.getUsername() + "'s account has been created.");
+        addRoleToUser(newUser.getUsername(), ROLES.ROLE_USER.toString());
+        setDefaultCosmetic(newUser.getUsername());
+        return newUser;
+    }
+
+    @Override
+    public User getUser(String username) {
+        log.info("Fetching user " + username);
+        return userRepository.findByUsername(username);
+    }
+
+    private boolean usernameAlreadyExists(String username) {
+        return userRepository.findByUsername(username) != null;
+    }
+
+    private boolean emailAlreadyExists(String email) {
+        return userRepository.findByEmail(email) != null;
+    }
+
+    @Override
+    public void addRoleToUser(String username, String roleName) throws UsernameNotFoundException, InvalidRoleException {
```

Comment:
We don't need to do this because with the current impl, all new users have the ROLE_USER role. For now we don't make any use of the ROLE_ADMIN role. We just have it here incase we want to make use of it in the future. So I'd say for now, it's okay.